atlasdb-impl-shared/src/main/java/com/palantir/atlasdb/keyvalue/api/cache/CacheMetrics.java

```diff
@@ -0,0 +1,96 @@
+    private final AccumulatingValueMetric hits;
+    private final AccumulatingValueMetric misses;
+    private final AccumulatingValueMetric cacheSize;
+    private final CurrentValueMetric<Integer> eventCacheValidationFailures;
+    private final CurrentValueMetric<Integer> valueCacheValidationFailures;
+    private final MetricsManager metricsManager;
+
+    private CacheMetrics(
+            AccumulatingValueMetric hits,
+            AccumulatingValueMetric misses,
+            AccumulatingValueMetric cacheSize,
+            CurrentValueMetric<Integer> eventCacheValidationFailures,
+            CurrentValueMetric<Integer> valueCacheValidationFailures,
+            MetricsManager metricsManager) {
+        this.hits = hits;
+        this.misses = misses;
+        this.cacheSize = cacheSize;
+        this.eventCacheValidationFailures = eventCacheValidationFailures;
+        this.valueCacheValidationFailures = valueCacheValidationFailures;
+        this.metricsManager = metricsManager;
+    }
+
+    public static CacheMetrics create(MetricsManager metricsManager) {
+        return new CacheMetrics(
+                metricsManager.registerOrGetGauge(
+                        CacheMetrics.class, AtlasDbMetricNames.CACHE_HITS, AccumulatingValueMetric::new),
+                metricsManager.registerOrGetGauge(
+                        CacheMetrics.class, AtlasDbMetricNames.CACHE_MISSES, AccumulatingValueMetric::new),
+                metricsManager.registerOrGetGauge(
+                        CacheMetrics.class, AtlasDbMetricNames.CACHE_SIZE, AccumulatingValueMetric::new),
+                metricsManager.registerOrGetGauge(
+                        CacheMetrics.class,
+                        AtlasDbMetricNames.EVENT_CACHE_VALIDATION_FAILURES,
+                        CurrentValueMetric::new),
+                metricsManager.registerOrGetGauge(
+                        CacheMetrics.class,
+                        AtlasDbMetricNames.VALUE_CACHE_VALIDATION_FAILURES,
+                        CurrentValueMetric::new),
+                metricsManager);
+    }
+
+    public void registerHits(long number) {
+        hits.accumulateValue(number);
+    }
+
+    public void registerMisses(long number) {
+        misses.accumulateValue(number);
+    }
+
+    public void increaseCacheSize(long added) {
+        cacheSize.accumulateValue(added);
+    }
+
+    public void decreaseCacheSize(long removed) {
+        cacheSize.accumulateValue(-removed);
+    }
+
+    public void registerEventCacheValidationFailure() {
+        eventCacheValidationFailures.setValue(1);
+    }
+
+    public void registerValueCacheValidationFailure() {
+        valueCacheValidationFailures.setValue(1);
+    }
+
+    public void setMaximumCacheSize(long maximumCacheSize) {
+        metricsManager.registerOrGetGauge(
+                CacheMetrics.class,
+                AtlasDbMetricNames.CACHE_RATIO_USED,
+                () -> () -> cacheSize.getValue() / (double) maximumCacheSize);
```

Comment:
sanity check: This method is only called once, right? If it's not then there may be some unexpected consequences (the `MetricSupplier` isn't invoked if the metric already exists).

Reply:
It should only be called once per TM, so we should be fine. I'm fine seeing if it breaks and fixing if we need to